File: examples/python/dex_examples/patterns/hash-partitioned-attribute-map/customer_directory_flow.py
# Copyright (c) 2022-2026 Super Durable, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from dataclasses import dataclass

from dex import AttributeMap, Context, Flow, PersistenceSchema, RPCResult, StepList, rpc

FLOW_ID = "customer-directory"
PARTITION_COUNT = 1000
FNV_OFFSET_BASIS_32 = 2166136261
FNV_PRIME_32 = 16777619


@dataclass
class CustomerProfile:
    email_address: str
    full_name: str
    company_name: str
    customer_tier: str


@dataclass
class CustomerProfilePartition:
    profiles_by_canonical_email: dict[str, CustomerProfile]


class CustomerDirectoryFlow(Flow[None]):
    customer_profiles_by_email_partition = AttributeMap(
        "customer_profiles_by_email_partition",
        CustomerProfilePartition,
    )

    def get_steps(self) -> StepList[None]:
        return StepList.empty()

    def get_persistence_schema(self) -> PersistenceSchema:
        return PersistenceSchema.of(self.customer_profiles_by_email_partition)

    @rpc(name="UpsertCustomerProfile")
    def upsert_customer_profile(
        self,
        context: Context,
        input: CustomerProfile,
    ) -> RPCResult[CustomerProfile]:
        canonical_email, partition_name, _ = email_partition(input.email_address)
        input.email_address = canonical_email
        try:
            partition = self.customer_profiles_by_email_partition.get(
                context, partition_name
            )
        except KeyError:
            partition = CustomerProfilePartition({})
        partition.profiles_by_canonical_email[canonical_email] = input
        self.customer_profiles_by_email_partition.set(
            context, partition_name, partition
        )
        return RPCResult(input)

    @rpc(name="GetCustomerProfileByEmail")
    def get_customer_profile_by_email(
        self,
        context: Context,
        input: str,
    ) -> RPCResult[CustomerProfile]:
        canonical_email, partition_name, _ = email_partition(input)
        try:
            partition = self.customer_profiles_by_email_partition.get(
                context, partition_name
            )
            profile = partition.profiles_by_canonical_email[canonical_email]
        except KeyError as error:
            raise ValueError(
                f'customer profile "{canonical_email}" not found'
            ) from error
        return RPCResult(profile)


def email_partition(email_address: str) -> tuple[str, str, int]:
    canonical_email = canonical_email_address(email_address)
    hash_value = fnv1a_32(canonical_email.encode("ascii"))
    return (
        canonical_email,
        f"partition-{hash_value % PARTITION_COUNT:03d}",
        hash_value,
    )


def canonical_email_address(email_address: str) -> str:
    try:
        encoded = email_address.encode("ascii")
    except UnicodeEncodeError as error:
        raise ValueError("emailAddress must contain only ASCII characters") from error
    start = 0
    while start < len(encoded) and _is_ascii_whitespace(encoded[start]):
        start += 1
    end = len(encoded)
    while end > start and _is_ascii_whitespace(encoded[end - 1]):
        end -= 1
    if start == end:
        raise ValueError("emailAddress is required")
    canonical = bytearray(encoded[start:end])
    for index, value in enumerate(canonical):
        if ord("A") <= value <= ord("Z"):
            canonical[index] = value + ord("a") - ord("A")
    return canonical.decode("ascii")


def fnv1a_32(value: bytes) -> int:
    hash_value = FNV_OFFSET_BASIS_32
    for octet in value:
        hash_value ^= octet
        hash_value = (hash_value * FNV_PRIME_32) & 0xFFFFFFFF
    return hash_value


def _is_ascii_whitespace(value: int) -> bool:
    return value == ord(" ") or ord("\t") <= value <= ord("\r")

File: examples/java/src/main/java/io/superdurable/dex/patterns/hashpartitionedattributemap/CustomerDirectoryFlow.java
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.superdurable.dex.patterns.hashpartitionedattributemap;

import io.superdurable.dex.AttributeMap;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.RPC;
import io.superdurable.dex.RPCResult;
import io.superdurable.dex.StepList;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import org.springframework.stereotype.Component;

@Component
public class CustomerDirectoryFlow implements Flow<Void> {
    public static final String FLOW_ID = "customer-directory";
    public static final int PARTITION_COUNT = 1000;
    private static final long FNV_OFFSET_BASIS_32 = 2166136261L;
    private static final long FNV_PRIME_32 = 16777619L;

    public final AttributeMap<CustomerProfilePartition> customerProfilesByEmailPartition =
            AttributeMap.define(
                    "customer_profiles_by_email_partition",
                    CustomerProfilePartition.class);

    @Override
    public StepList<Void> getSteps() {
        return StepList.empty();
    }

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(customerProfilesByEmailPartition);
    }

    @RPC(name = "UpsertCustomerProfile")
    public RPCResult<CustomerProfile> upsertCustomerProfile(
            final Context context,
            final CustomerProfile input) {
        final EmailPartition partition = emailPartition(input.emailAddress);
        input.emailAddress = partition.canonicalEmail;
        CustomerProfilePartition profiles = customerProfilesByEmailPartition.get(
                context,
                partition.partitionName);
        if (profiles == null) {
            profiles = new CustomerProfilePartition(new HashMap<String, CustomerProfile>());
        }
        if (profiles.profilesByCanonicalEmail == null) {
            profiles.profilesByCanonicalEmail = new HashMap<String, CustomerProfile>();
        }
        profiles.profilesByCanonicalEmail.put(partition.canonicalEmail, input);
        customerProfilesByEmailPartition.set(context, partition.partitionName, profiles);
        return RPCResult.of(input);
    }

    @RPC(name = "GetCustomerProfileByEmail")
    public RPCResult<CustomerProfile> getCustomerProfileByEmail(
            final Context context,
            final String emailAddress) {
        final EmailPartition partition = emailPartition(emailAddress);
        final CustomerProfilePartition profiles = customerProfilesByEmailPartition.get(
                context,
                partition.partitionName);
        final CustomerProfile profile = profiles == null
                ? null
                : profiles.profilesByCanonicalEmail.get(partition.canonicalEmail);
        if (profile == null) {
            throw new IllegalArgumentException(
                    "customer profile \"" + partition.canonicalEmail + "\" not found");
        }
        return RPCResult.of(profile);
    }

    public static EmailPartition emailPartition(final String emailAddress) {
        final String canonicalEmail = canonicalEmailAddress(emailAddress);
        final long hash = fnv1a32(canonicalEmail.getBytes(StandardCharsets.US_ASCII));
        return new EmailPartition(
                canonicalEmail,
                hash,
                String.format("partition-%03d", hash % PARTITION_COUNT));
    }

    public static String canonicalEmailAddress(final String emailAddress) {
        if (emailAddress == null) {
            throw new IllegalArgumentException("emailAddress is required");
        }
        int start = 0;
        int end = emailAddress.length();
        for (int index = 0; index < emailAddress.length(); index++) {
            if (emailAddress.charAt(index) > 0x7f) {
                throw new IllegalArgumentException(
                        "emailAddress must contain only ASCII characters");
            }
        }
        while (start < end && isAsciiWhitespace(emailAddress.charAt(start))) {
            start++;
        }
        while (end > start && isAsciiWhitespace(emailAddress.charAt(end - 1))) {
            end--;
        }
        if (start == end) {
            throw new IllegalArgumentException("emailAddress is required");
        }
        final char[] canonical = emailAddress.substring(start, end).toCharArray();
        for (int index = 0; index < canonical.length; index++) {
            if (canonical[index] >= 'A' && canonical[index] <= 'Z') {
                canonical[index] = (char) (canonical[index] + ('a' - 'A'));
            }
        }
        return new String(canonical);
    }

    public static long fnv1a32(final byte[] value) {
        long hash = FNV_OFFSET_BASIS_32;
        for (byte octet : value) {
            hash ^= octet & 0xff;
            hash = (hash * FNV_PRIME_32) & 0xffffffffL;
        }
        return hash;
    }

    private static boolean isAsciiWhitespace(final char value) {
        return value == ' ' || (value >= '\t' && value <= '\r');
    }

    public static class CustomerProfile {
        public String emailAddress;
        public String fullName;
        public String companyName;
        public String customerTier;

        public CustomerProfile() {
        }

        public CustomerProfile(
                final String emailAddress,
                final String fullName,
                final String companyName,
                final String customerTier) {
            this.emailAddress = emailAddress;
            this.fullName = fullName;
            this.companyName = companyName;
            this.customerTier = customerTier;
        }
    }

    public static class CustomerProfilePartition {
        public Map<String, CustomerProfile> profilesByCanonicalEmail;

        public CustomerProfilePartition() {
        }

        public CustomerProfilePartition(
                final Map<String, CustomerProfile> profilesByCanonicalEmail) {
            this.profilesByCanonicalEmail = profilesByCanonicalEmail;
        }
    }

    public static class EmailPartition {
        public String canonicalEmail;
        public long hash;
        public String partitionName;

        public EmailPartition() {
        }

        public EmailPartition(
                final String canonicalEmail,
                final long hash,
                final String partitionName) {
            this.canonicalEmail = canonicalEmail;
            this.hash = hash;
            this.partitionName = partitionName;
        }
    }
}
